### ISSUE TITLE
[HOLD for payment 2023-11-29] [Redesign Unread Markers] Fix unread markers and migrate logic to ReportActionsList component

### ISSUE THREAD

**Issue by MonilBhavsar** [Expensify/App#23171]
Tracking issue https://github.com/Expensify/App/issues/15212

As mentioned in the doc, implement client side changes to update unread marker https://docs.google.com/document/d/144r92kq3lvJlT2sg9wQGzFHdq5PPyai4eCs5jsBWrZw/edit

1. Move the logic to handle unread marker from ReportActionsView to ReportActionsList component
2. While calling openReport command, use optimistic data if possible

--- 

### Root cause:

1. lastReadTime being updated by the server and optimistically, too early.

Currently, when fetching report through OpenReport API command, the client optimistically marks the message  as read or update the lastReadTime [here](https://github.com/Expensify/App/blob/0d821677c56b8f1ed479463501ad2ab90ee4412a/src/libs/actions/Report.js#L368), and the server also updates the lastReadTime and returns updated lastReadTime [here](https://github.com/Expensify/Web-Expensify/blob/0250f5f9a98e512a185de09ef18ce08058b39640/lib/ReportAPI.php#L4714-L4729).  This makes lastReadTime greater than created time, and since reportAction.created < report.lastReadTime,  we fail to render unread marker 

2.  Taking into consideration incorrect reportActions when calculating whether to display unread marker

Currently, when a user navigates to chat from LHN, or directly through URL, or from notification, we render ReportScreen component, and then render ReportActionsView component by passing reportActions as props [here](https://github.com/Expensify/App/blob/main/src/pages/home/ReportScreen.js#L328).
We calculate whether or where we should display an unread marker in the chat, when the chat is loaded initially in the constructor of the ReportActionsView component [here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsView.js#L72). This calculation is based on the reportActions(messages) associated with that chat.
When switching chats, we observed that there is a slight delay in getting reportActions of new chat or the chat one is switching to, and it seems like the first couple of renders happen with the reportActions one is switching from. As a result, we incorrectly calculate unread marker based on the reportActions of the previous chat.

3. Some Performance concerns

We calculate where we should display unread marker in the ReportUtils.getNewMarkerReportActionID function [here](https://github.com/Expensify/App/blob/1ab36df9fa03a3e85cb68c64de71eca3d559a3dd/src/libs/ReportUtils.js#L2008). 
This function is called multiple times - one from the constructor of the ReportActionsView component and multiple times from the componentDidUpdate equivalent method  in the ReportActionsView component. The function has complexity of O(n), and it is being performed during component re-rendering.

### Proposed solution:

![UnreadMarker](https://github.com/Expensify/App/assets/32012005/f609b4db-12fd-4623-9588-72a4831835de)

Right now, the whole logic to render new markers is in the ReportActionsView component. We will remove that logic from the ReportActionsView component and move it to the  [ReportActionsList](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsList.js)
 component, where we render reportActions(messages).

#### Removing logic from ReportActionsView
From the [ReportActionsView](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsView.js) we are removing the newMarkerReportActionID from the state.
Also we are removing some functions to handle and update that marker, ReportUtils.isUnread function [here](https://github.com/Expensify/App/blob/1ab36df9fa03a3e85cb68c64de71eca3d559a3dd/src/libs/ReportUtils.js#L1765)  and ReportUtils.getNewMarkerReportActionID function [here](https://github.com/Expensify/App/blob/0ee7dfa0dadb6ba11ca4cf08085df9003b0fb452/src/libs/ReportUtils.js#L2143). Additionally, any logic that uses them will be removed too.

#### Handling unread marker
Now that we’re using hooks, we need to have an indicator to check if a user is active on a chat or not. This is because if the user is active on chat and receives messages in real time, then we don’t want to display unread marker.
To accomplish this, we will utilize a ref userActiveSince  that stores the timestamp when users enter the chat.

```js
useEffect(() => {
   userActiveSince.current = DateUtils.getDBTime();
}, []);
```
The logic to know if and where we should display an unread marker will be calculated into the renderItem function in the ReportActionsList component [here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsList.js), while the items in the list are being created.

We display an unread marker if a message was created(reportAction.created) after the user last read the chat(report.lastReadTime).The basic function to check if we should display an unread message will be as below:

```js
function isUnreadMessage(message, lastRead) {
    return Boolean(message.created && lastRead < message.created);
}
```
To determine where we should display an unread marker, we check if the current message is read and the next message is unread. If such is the case, we will display an unread marker there. 
The logic to determine the position of unread marker is as below:

```ks
const nextMessage = sortedReportActions[index + 1];

const isCurrentMessageUnread = isUnreadMessage(reportAction, report.lastReadTime);

const shouldDisplayNewMarker = isCurrentMessageUnread && !isUnreadMessage(nextMessage, report.lastReadTime);
```

To keep track of when a green marker is being displayed, we will store the information in a ref called currentUnreadMarker that holds reportActionID above which we should display unread marker.  If the marker is already rendered and we have  information about the marker, we will reuse it.

```js
if (!currentUnreadMarker.reportActionIDcurrent) {
    // ...

	if (shouldDisplayNewMarker && canDisplayMarker) {
   		currentUnreadMarker.reportActionIDcurrent = reportAction.reportActionID;
     }
} else {
     shouldDisplayNewMarker = reportAction.reportActionID === currentUnreadMarker.reportActionIDcurrent;
}
```
#### Marking messages as read explicitly

Since the server will no longer update the report’s lastReadTime data in OpenReport API command, we need to mark the report as read from the client's side in all cases. After receiving reportActions, we will inform the server to update that lastReadTime by calling [Report.readNewestAction](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js). This new function will invoke the ReadNewestAction API command, passing the reportID and the lastReadTime to be set.

There are some cases to take into consideration:
We first check if there is an unread message using the [ReportUtils.isUnread](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L1815)
We would need to take into consideration the scrolling position. If the user is scrolled up, that means they have’t read the message and we won’t mark it as read.

```js
useEffect(() => {
    if (ReportUtils.isUnread(report)) {
        if (scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
            Report.readNewestAction(report.reportID);
    	   } else { ...
```

#### Marking message as unread manually

The user has the ability to mark any message as unread, which is achieved by modifying the lastReadTime in markCommentAsUnread function [here](https://github.com/Expensify/App/blob/0d821677c56b8f1ed479463501ad2ab90ee4412a/src/libs/actions/Report.js#L698). For manually marking message as unread, we subtract 1 millisecond from the lastReadTime.
To determine if a user has manually marked a message as unread, we check if the lastReadTime is less than the current time using DateUtils.getDBTime() and if there is an unread message using [ReportUtils.isUnread](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L1815). If this condition is true, we clear the currentUnreadMarker to ensure the unread marker can be re-calculated, and update the messageManuallyMarkedAsUnread flag.

Also, we don’t need to display an unread marker on the sender’s side when they are sending a new message, unless they manually marked the message as unread.
We’ll have the following check to ensure this behavior:

```js
if (!messageManuallyMarkedAsUnread) {
   shouldDisplayNewMarker = shouldDisplayNewMarker && reportAction.actorAccountID !== Report.getCurrentUserAccountID();
}
```

#### Scrolling behaviors

Currently logic to track scrolling is handled in the ReportActionsView component [here](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsView.js#L278-L296). 
We will be transferring the scrolling tracking functionality from ReportActionsView to ReportActionsList and reuse it. 

Additionally, we display a floating element when the user is scrolled up and there are new latest messages to read. It is rendered using the FloatingMessageCounter in ReportActionsView component. We’ll also move it to ReportActionsList component and its visibility(whether we should display floating element or not) will be managed using a useState hook, which will be calculated based on the user's scrolling behavior.

We also need logic to determine if we should display an unread marker or floating element considering the user's position in the chat.  If the user has scrolled a little bit and the latest messages are visible then we display an unread marker, else if the user has scrolled much and messages are not visible then we display a floating element to allow the user to jump to the latest messages. 
A constant canDisplayMarker will handle this behavior as described below

```js
const canDisplayMarker = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < userActiveSince.current : true;
```

### Fixing delayed rendering issue

When switching between Reports, the ReportScreen component passes old reportActions from Onyx, during initial renders which can lead to incorrect calculation of the unread marker. To address this issue and avoid any unusual loading delays when fetching the reportActions. We will update [Report.openReport](https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js#L361) to check if we already have the reportActions data, we use an empty object in the optimisticData to prevent unnecessary re-renders and ensure that the correct data is passed. However, if the reportActions data is not available, we handle it as normal and fetch the data as needed.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c3eb8d200e91874e</li>
        <li>Upwork Job ID: 1681647663368384512</li>
        <li>Last Price Increase: 2023-07-19</li>
    </ul>
</details>

**Comment by MonilBhavsar** [comment]
@mananjadhav I have updated OP with problem/solution format and with all the details required for this issue. Please let me know if that sounds good

**Comment by mananjadhav** [comment]
Thanks for the details. I'll go through it.

**Comment by trjExpensify** [comment]
Did that help clarify @mananjadhav?

**Comment by MonilBhavsar** [comment]
@gedu is out for the week, so I'll be working on this in parallel

**Comment by MonilBhavsar** [comment]
Testing the PR's

**Comment by trjExpensify** [comment]
How's the PR testing going?

**Comment by MonilBhavsar** [comment]
Going good!
We need to fix scrolling behavior. @gedu is back from ooo and will be working on this from today.

**Comment by mananjadhav** [comment]
Won't be able to get this soon. Asked to take over [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1690948784218259).

**Comment by trjExpensify** [comment]
👋 what's the latest here? Which PR from @gedu is linked to this issue?

**Comment by trjExpensify** [comment]
Thanks!

**Comment by trjExpensify** [comment]
I'm putting this back on `weekly`, but @allroundexperts @gedu @MonilBhavsar can you provide an update on what's happening with this issue and all these deploy blockers being linked?

**Comment by MonilBhavsar** [comment]
There were deploy blockers that were sussing this PR, but turned out to be not. Also, it's a while since the PR was deployed for this implementation issue.

It went well! One case to be fixed is for Delete Message case @gedu is working on it

**Comment by trjExpensify** [comment]
@gedu @MonilBhavsar mind providing an update on where we're at and the next steps for this issue?

**Comment by MonilBhavsar** [comment]
Here's the latest https://github.com/Expensify/App/pull/25935#issuecomment-1775268982

**Comment by trjExpensify** [comment]
Okay, conflicts are resolved so @allroundexperts let's get it reviewed before they strike again! 😅

**Comment by trjExpensify** [comment]
Perfect, thank you!

**Comment by trjExpensify** [comment]
Putting this back on `weekly`. @allroundexperts what's the latest?

**Comment by situchan** [comment]
Potential regressions from https://github.com/Expensify/App/pull/25935:
- https://github.com/Expensify/App/issues/31617
- https://github.com/Expensify/App/issues/31621
- https://github.com/Expensify/App/issues/31623

**Comment by trjExpensify** [comment]
Sure done! Can I get an update on progress please? :)

**Comment by shubham1206agra** [comment]
@trjExpensify The PR is in production.
But we need some refinement/improvement we can do with follow-up PR.

**Comment by shubham1206agra** [comment]
The follow up PR can fix these issues
- https://github.com/Expensify/App/issues/31714
- https://github.com/Expensify/App/issues/31617
- https://github.com/Expensify/App/issues/31621

**Comment by trjExpensify** [comment]
Oh nice, so we'll be looking to close those after this regression period?

**Comment by situchan** [comment]
There was regression - https://github.com/Expensify/App/issues/31623 which was deploy blocker

**Comment by shubham1206agra** [comment]
@trjExpensify This was an edge case which may have been missed during testing

**Comment by trjExpensify** [comment]
Sorry, I'm not sure I'm following?

**Comment by shubham1206agra** [comment]
I wanted to say that I thought I checked the new chat behaviour by creating a thread. But it seems another problem occurred with chatting with new users.

**Comment by shubham1206agra** [comment]
@trjExpensify Should we create a follow-up PR for fixing https://github.com/Expensify/App/issues/23171#issuecomment-1822983891 or we let them fix in the issues respectively?

**Comment by trjExpensify** [comment]
Yes please, you should if they are regressions from your PR. There's no point having three different issues fixing the same root cause.

**Comment by shubham1206agra** [comment]
@gedu Can you spin up a follow-up PR then?

**Comment by gedu** [comment]
I will review it too, I was checking other possible bug

**Comment by shubham1206agra** [comment]
I think @gedu, you need to fix these issues in another PR.
Maybe first check if these issues comes from this. By reverting PR internally and then try to repro again.

**Comment by trjExpensify** [comment]
How we getting on here?

**Comment by shubham1206agra** [comment]
Waiting for @gedu to raise a PR for the above issues

**Comment by trjExpensify** [comment]
Cool, @gedu do you have an ETA on that?

**Comment by trjExpensify** [comment]
Reached out to @gedu.

**Comment by gedu** [comment]
With Monil we are looking into the bugs:
- https://github.com/Expensify/App/issues/31714#issuecomment-1842847853
- https://github.com/Expensify/App/issues/31621#issuecomment-1849825994

**Comment by trjExpensify** [comment]
Okay, so @shubham1206agra what exactly do you think we're waiting on?

**Comment by shubham1206agra** [comment]
Another PR to fix these problems

**Comment by trjExpensify** [comment]
What problems and how are they different or not being addressed in these issues @gedu linked? https://github.com/Expensify/App/issues/23171#issuecomment-1849853728

**Comment by shubham1206agra** [comment]
I mean above issues only

**Comment by shubham1206agra** [comment]
@MonilBhavsar and @gedu seems to be working on a fix

**Comment by trjExpensify** [comment]
Cool, so why is this issue still open if we have the others?

**Comment by shubham1206agra** [comment]
Don't know. We can close this if you want after the payment.
I will fill the checklist here later.

**Comment by trjExpensify** [comment]
Awaiting the checklist!

**Comment by shubham1206agra** [comment]
@trjExpensify I don't think we need checklist as confirmed by @allroundexperts. You can pay and close this issue here.

**Comment by trjExpensify** [comment]
Where was that confirmed, sorry?

**Comment by allroundexperts** [comment]
@trjExpensify I think @shubham1206agra mean't that this was a redesign / refactor and the checklist is not applicable here.

**Comment by trjExpensify** [comment]
Hm gotcha, so @MonilBhavsar I take it with the project you had a bunch of manual tests to handover to Applause to update/add/remove based on the changes, and you'll do that as part of the parent issue of the project?

In terms of payment then guys, can we confirm how many regressions came from this issue to tally up what's due? 

P.S as a heads up, tomorrow is my last day for the holidays and I'll be back on Jan 3rd. So let me know tomorrow and I can settle up ahead of then. :)

**Comment by MonilBhavsar** [comment]
@trjExpensify yes, we add tests for all cases together https://docs.google.com/document/d/1j1UKjl_zVmS_EVVtpeH2ugrlt6INuWKlnp3X_o_WXRY/edit#bookmark=id.v1lsuitewwbd

There were couple of small regressions from this mega refactor PR https://github.com/Expensify/App/pull/18637/files
Given it refactored base ReportActionsList component and we tried our best to avoid them, I think we should pay full amount to @allroundexperts for reviewing it

**Comment by shubham1206agra** [comment]
@MonilBhavsar me too :)

**Comment by MonilBhavsar** [comment]
Oh right 👍  @shubham1206agra finished the review and checklist and @allroundexperts reviewed initially.
@trjExpensify do you think we should split the payment here?

**Comment by trjExpensify** [comment]
>@trjExpensify yes, we add tests for all cases together https://docs.google.com/document/d/1j1UKjl_zVmS_EVVtpeH2ugrlt6INuWKlnp3X_o_WXRY/edit#bookmark=id.v1lsuitewwbd

Has this been handed off to Applause to add to TestRail? (Can't access the doc btw).

**Comment by trjExpensify** [comment]
>@trjExpensify do you think we should split the payment here?

Yeah, makes sense. I'm assuming this was at the old $1k bounty level, so this is $500 each? Once confirmed, I'll send the offers.

**Comment by trjExpensify** [comment]
Let me know on the above!

**Comment by trjExpensify** [comment]
Same, Melv.

**Comment by MonilBhavsar** [comment]
> > @trjExpensify do you think we should split the payment here?
> 
> Yeah, makes sense. I'm assuming this was at the old $1k bounty level, so this is $500 each? Once confirmed, I'll send the offers.

Sounds good and fair to me 👍

**Comment by trjExpensify** [comment]
Cool, offer has been sent to @shubham1206agra. @allroundexperts, you can request, right?

**Comment by allroundexperts** [comment]
Yep, requested. Feel free to close. Thanks!

**Comment by trjExpensify** [comment]
Still need @shubham1206agra to accept :)

**Comment by shubham1206agra** [comment]
Accepted

**Comment by trjExpensify** [comment]
Paid!

**Comment by JmillsExpensify** [comment]
$500 payment approved for @allroundexperts based on [this comment](https://github.com/Expensify/App/issues/23171#issuecomment-1875133077).
